src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString(attribute_, values_);
+    }
+
+  private:
+    std::string attribute_;
+    std::vector<T> values_;
+};
+
+class NodeSetBasicPopulation: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicPopulation(std::string name)
+        : name_(std::move(name)) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        if (name_ == np.name()) {
+            return np.selectAll();
+        }
+
+        return Selection{{}};
+    }
+
+    std::string toJSON() const final {
+        const std::vector<std::string> values = {name_};
+        return toString("population", values);
+    }
+
+  private:
+    std::string name_;
+};
+
+class NodeSetBasicNodeIds: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicNodeIds(Selection::Values&& values)
+        : values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */,
+                          const NodePopulation& /* unused */) const final {
+        return Selection::fromValues(values_.begin(), values_.end());
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", values_);
+    }
+
+  private:
+    Selection::Values values_;
+};
+
+using CompoundTargets = std::vector<std::string>;
+class NodeSetCompoundRule: public NodeSetRule
+{
+  public:
+    NodeSetCompoundRule(std::string attribute, const CompoundTargets& targets)
+        : attribute_(std::move(attribute))
+        , targets_(targets) {}
+
+    Selection materialize(const NodeSets& ns, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& target : targets_) {
+            ret = ret | ns.materialize(target, np);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", targets_);
+    }
+
+  private:
+    std::string attribute_;
+    CompoundTargets targets_;
+};
+
+int64_t get_integer_or_throw(const json& el) {
+    auto v = el.get<double>();
+    if (std::floor(v) != v) {
+        throw SonataError("Only allowed integers in node set rules");
+    }
+    return static_cast<int64_t>(v);
+}
+
+NodeSetRules _dispatch_node(const json& contents) {
+    NodeSetRules ret;
+
+    for (auto& el : contents.items()) {
+        const std::string& attribute = el.key();
+        if (el.value().is_number()) {
+            if (attribute == "population") {
+                throw SonataError("'population' must be a string");
+            }
+
+            int64_t v = get_integer_or_throw(el.value());
+            if (attribute == "node_id") {
+                if (v < 0) {
+                    throw SonataError("'node_id' must be positive");
+                }
+
+                ret.emplace_back(new NodeSetBasicNodeIds({static_cast<Selection::Value>(v)}));
+            } else {
+                ret.emplace_back(new NodeSetBasicRule<int64_t>(attribute, {v}));
```

Comment:
is the <int64_t> ok if the parameter inside the is let say a int32 or uint32 ?

Reply:
should be, it's a widening, which is safe, afaik.

Reply:
but not for the uint64, no ?

Reply:
I thought you were just asking about `int32`; yes, it would fail for `uint64` greater than 4294967296;  I'm not sure what the right behavior is in that case.

Reply:
Yeah did not choose my example correctly ...

I think we can live with the uint64 pb for the moment. 
Just leave a comment so if one day this appears to be a problem we know where to look at